admin/cluster_setup.py

```diff
@@ -0,0 +1,497 @@
+    ]
+
+    synopsis = ('Usage: cluster-setup --distribution <distribution> '
+                '[--provider <provider>]')
+
+    def __init__(self, top_level):
+        """
+        :param FilePath top_level: The top-level of the flocker repository.
+        """
+        Options.__init__(self)
+        self.docs['provider'] = self.docs['provider'].format(
+            self._get_provider_names()
+        )
+        self.top_level = top_level
+        self['variants'] = []
+
+    def _get_provider_names(self):
+        """
+        Find the names of all supported "providers" (eg Vagrant, Rackspace).
+
+        :return: A ``list`` of ``str`` giving all such names.
+        """
+        return prefixedMethodNames(self.__class__, "_runner_")
+
+    def opt_variant(self, arg):
+        """
+        Specify a variant of the provisioning to run.
+
+        Supported variants: distro-testing, docker-head, zfs-testing.
+        """
+        self['variants'].append(Variants.lookupByValue(arg))
+
+    def dataset_backend_configuration(self):
+        """
+        Get the configuration corresponding to storage driver chosen by the
+        command line options.
+        """
+        drivers = self['config'].get('storage-drivers', {})
+        configuration = drivers.get(self['dataset-backend'], {})
+        return configuration
+
+    def dataset_backend(self):
+        """
+        Get the storage driver the testing nodes will use.
+
+        :return: A constant from ``DatasetBackend`` matching the name of the
+            backend chosen by the command-line options.
+        """
+        configuration = self.dataset_backend_configuration()
+        # Avoid requiring repetition of the backend name when it is the same as
+        # the name of the configuration section.  But allow it so that there
+        # can be "great-openstack-provider" and "better-openstack-provider"
+        # sections side-by-side that both use "openstack" backend but configure
+        # it slightly differently.
+        dataset_backend_name = configuration.get(
+            "backend", self["dataset-backend"]
+        )
+        try:
+            return DatasetBackend.lookupByName(dataset_backend_name)
+        except ValueError:
+            raise UsageError(
+                "Unknown dataset backend: {}".format(
+                    dataset_backend_name
+                )
+            )
+
+    def postOptions(self):
+        if self['distribution'] is None:
+            raise UsageError("Distribution required.")
+
+        if self['config-file'] is not None:
+            config_file = FilePath(self['config-file'])
+            self['config'] = yaml.safe_load(config_file.getContent())
+        else:
+            self['config'] = {}
+
+        if self['app-template'] is not None:
+            template_file = FilePath(self['app-template'])
+            self['template'] = yaml.safe_load(template_file.getContent())
+        else:
+            self['template'] = {}
```

Comment:
If `app-template` is not set, but `apps-per-node > 0` the code gives the error:
```
Application 'app_172.31.37.42_0' has a config error. Missing 'image' key.
```

Raise a UsageError if this combination of options is provided.